torch/optim/rprop.py

```diff
@@ -247,17 +284,26 @@ def _single_tensor_rprop(
             sign = grad.mul(prev.clone()).sign()
         else:
             sign = grad.mul(prev).sign()
-        sign[sign.gt(0)] = etaplus
-        sign[sign.lt(0)] = etaminus
-        sign[sign.eq(0)] = 1
+
+        if capturable:
+            sign.copy_(torch.where(sign.gt(0), etaplus, sign))
+            sign.copy_(torch.where(sign.lt(0), etaminus, sign))
+            sign.copy_(torch.where(sign.eq(0), 1, sign))
```

Comment:
Not expected to be in the scope of the PR but I wonder if there's a fast way to assign all of these. Previously the index select would be pretty fast and only write into N total elements for all three lines. Now, we write into 3N total elements, in addition to calling `torch.where`. Would be interesting to have an API that would call the same lambda on every element in the Tensor to copy into, like a:
```
sign.copy_(lambda x: etaplus if x > 0 else (etaminus if x < 0 else 1))
```
This API wouldn't be general enough to replace the grad.copy_ line later though :/

Reply:
if we had https://github.com/pytorch/pytorch/issues/117884 we could at least do this in two kernels which would be nice. We could possibly implement what you're talking about with a higher order op. Luckily though for compiling this doesn't appear to affect the kernel count, we still generate an optimized kernel